File: AlgoExpert/2. Medium/C++/Monotonic Array/Monotonic Array [O(n)].cpp
#include <iostream>

using namespace std;

bool breaksDirection(int direction, int previousInt, int currentInt) {
	int difference = currentInt - previousInt;
	if (direction > 0)
	return difference < 0;
	return difference > 0;
}

// O(n) time | O(1) space - where n is the length of the array

bool isMonotonic(vector<int> array) {
	if (array.size() <= 2)
		return true;
	int direction = array[1] - array[0];
	for (int i = 2; i < array.size(); i++) {
		if (direction == 0) {
			direction = array[i] - array[i - 1];
			continue;
		}
		if (breaksDirection(direction, array[i - 1], array[i]))
			return false;
	}
	return true;
}

File: AlgoExpert/2. Medium/Python/Monotonic Array/Monotonic Array [O(n)].py
# O(n) time | O(1) space - where n is the length of the array

def isMonotonic(array):
	isNonDecreasing = True
	isNonIncreasing = True
	for i in range(1, len(array)):
		if array[i] < array[i - 1]:
			isNonDecreasing = False
		if array[i] > array[i - 1]:
			isNonIncreasing = False
	
	return isNonDecreasing or isNonIncreasing